src/shopify-cart.ts

```diff
@@ -0,0 +1,85 @@
+import { fetchVariantsPreorderState } from "./api";
+import { Cart, CartItem, PreorderAttributes } from "./cart";
+import { JSONObject } from "./interceptors";
+
+export type ShopifyCartItem = CartItem & {
+  id?: string;
+  quantity?: number;
+  properties?: Record<string, string>;
+};
+
+const ShopifyCart: Cart<ShopifyCartItem> = {
+  hasPreorderAttributes(item: ShopifyCartItem): boolean {
+    return !!item.properties?.["__releaseId"];
+  },
+
+  addPreorderAttributes(item: ShopifyCartItem, attrs: PreorderAttributes) {
+    return {
+      id: item.id,
+      variantId: item.variantId,
+      properties: {
+        ...item.properties,
+        __releaseId: attrs.releaseId,
+        "Purple Dot Pre-order": attrs.displayShipDates,
+      },
+    };
+  },
+
+  removePreorderAttributes(item: ShopifyCartItem) {
+    return {
+      id: item.id,
+      variantId: item.variantId,
+      properties: Object.fromEntries(
+        Object.entries(item.properties ?? {}).filter(
+          ([key]) => key !== "__releaseId" && key !== "Purple Dot Pre-order",
+        ),
+      ),
+    };
+  },
+
+  async fetchItems() {
+    // TODO: Add the locale to the URL
+    const response = await fetch("/cart.js");
+    const data = await response.json();
+
+    return data.items.map((item: JSONObject) => {
+      return {
+        id: item.key ?? item.id ?? item.variant_id,
+        variantId: item.variant_id,
+        properties: item.properties,
+      };
+    });
+  },
+
+  async decrementQuantity(id: string) {
+    // TODO: Fetch the cart and decrement one of the matching line items
```

Comment:
I think you can also do that directly using the key?

Reply:
I couldn't find a documented way of doing this. Do you have some docs?

Reply:
Ah you can't, sorry - all you can do is update the quantity but can't do an atomic decrement.